happi/audit.py

```diff
@@ -0,0 +1,424 @@
+        return [self.create_arg(item, arg) for arg in item.args]
+
+    def validate_kwargs(self, item):
+        """
+        Validates the kwargs of an item
+        """
+        return dict((key, self.create_arg(item, val))
+                    for key, val in item.kwargs.items())
+
+    def create_arg(self, item, arg):
+        """
+        Function borrowed from loader to create
+        correctly typed arguments from happi information
+        """
+        if not isinstance(arg, str):
+            return arg
+        return fill_template(arg, item, enforce_type=True)
+
+    def get_device_class(self, item):
+        """
+        Validates device_class field
+        """
+        device_class = item.get('device_class')
+        device = item.get('name')
+
+        if not device_class:
+            logger.warning('Detected a None vlaue for %s. '
+                           'The device_class cannot be None', device)
+            return self.report_code.MISSING
+        else:
+            try:
+                mod, cls = device_class.rsplit('.', 1)
+            except (Exception, ValueError) as e:
+                logger.warning('Wrong device name format: %s for %s, %s',
+                               device_class, device, e)
+                return self.report_code.INVALID
+            else:
+                packages = device_class.rsplit('.')
+                package = packages[0]
+                self._all_devices.add(package)
+                self._all_items.append(item)
+                return self._all_devices
+
+    def validate_import_class(self, item=None):
+        """
+        Validate the device_class of an item
+        """
+        device_class = item.get('device_class')
+        mod = device_class
+        if not device_class:
+            return self.report_code.MISSING
+        if '.' not in device_class:
+            logger.warning('Device class invalid %s for item %s',
+                           device_class, item)
+            return self.report_code.INVALID
+        else:
+            mod = device_class.rsplit('.')[0]
+        device = item.get('name')
+        if mod in self._all_devices:
+            # has been tested agains PyPi and was found there
+            try:
+                import_class(device_class)
+                return self.report_code.SUCCESS
+            except ImportError as e:
+                logger.warning('For device: %s, %s. Either %s is '
+                               'misspelled or %s is not part of the '
+                               'python environment', device, e,
+                               device_class, device_class)
+                return (self.report_code.INVALID, self.report_code.MISSING)
+        elif mod and mod not in self._all_devices:
+            logger.warning('Provided wrong device/module name: %s. '
+                           'This module does not exist in PyPi', device_class)
+            return self.report_code.INVALID
+
+    def search_pip_package(self, package):
```

Comment:
This is a cool idea, but I don't think it has a place in this audit script. It should just be up to the user to figure out why it's bad if it doesn't import.

It's a bit heavy, requires an internet connection, spawns new processes and slows down the audit process, etc.
Oh, and sometimes the import name of a package and its package name are two different things, even.

Edit: Reading below, it looks like @ZLLentz requested this. I think we should discuss this further.

Reply:
Yeah it works fine, but it takes some time to check, so it does slow down the audit process. Also, because it requires internet it is not ideal - I agree with that. 
I think you make a good point that we can leave it up to the user to figure out why it's bad if it does not import - what do you think @ZLLentz ?

Reply:
It's not fully necessary to do the pip search. Maybe we were a little "too helpful" here in ways where we make the script cumbersome to run.

Reply:
It could be an option, given that you've done the work here @cristinasewell. Making it optional will further complicate the already pretty complicated logic here, though. Is pypi searching something we could perhaps shelve for the next audit version?

Reply:
I am a fan of removing this if that's ok :)